File: src/data/scripts/everyFrames/niko_MPC_temporarySatelliteFleetDespawner.java
package data.scripts.everyFrames;

import com.fs.starfarer.api.EveryFrameScriptWithCleanup;
import com.fs.starfarer.api.campaign.BattleAPI;
import com.fs.starfarer.api.campaign.CampaignFleetAPI;
import data.scripts.campaign.econ.conditions.defenseSatellite.niko_MPC_satelliteHandlerCore;
import data.utilities.*;

public class niko_MPC_temporarySatelliteFleetDespawner implements EveryFrameScriptWithCleanup {

    public CampaignFleetAPI fleet;
    public niko_MPC_satelliteHandlerCore handler;

    public boolean done = false;

    public double advanceTimeSinceStart = 0;
    public int graceRuns = 0;

    public niko_MPC_temporarySatelliteFleetDespawner(CampaignFleetAPI fleet, niko_MPC_satelliteHandlerCore handler) {
        this.fleet = fleet;
        this.handler = handler;

        init();
    }

    private void init() {
        if (fleet == null) {
            prepareForGarbageCollection();
        }
    }

    @Override
    public void cleanup() {
        prepareForGarbageCollection();
    }

    @Override
    public boolean isDone() {
        return done;
    }

    @Override
    public boolean runWhilePaused() {
        return false;
    }

    @Override
    public void advance(float amount) {
        advanceTimeSinceStart += amount;

        if (fleet.getBattle() == null && graceRuns <= 0) {
            getRidOfFleet(); //this despawns the fleet, which calls the listener, which calls prepareforgc. its weird
            return;
        }
        else {
            niko_MPC_satelliteBattleTracker tracker = niko_MPC_satelliteUtils.getSatelliteBattleTracker();
            BattleAPI battle = fleet.getBattle();
            if (battle != null && !tracker.areSatellitesInvolvedInBattle(battle, handler)) { // sanity
                tracker.associateSatellitesWithBattle(battle, handler, battle.pickSide(fleet));
            }
          /*  if (handler.fleetForPlayerDialog == fleet) {
                handler.fleetForPlayerDialog = null;
            } */
        }
        //graceRuns--;
    }

    public void prepareForGarbageCollection() {

        if (fleet != null) {
            niko_MPC_memoryUtils.deleteMemoryKey(fleet.getMemoryWithoutUpdate(), niko_MPC_ids.temporaryFleetDespawnerId);
            fleet.removeScript(this);
            fleet = null;
        }
        handler = null;
        done = true;
    }


    private void getRidOfFleet() {
        boolean vanish = (advanceTimeSinceStart < 1); //arbitrary number
        niko_MPC_fleetUtils.despawnSatelliteFleet(fleet, vanish);
    }
}

File: src/data/scripts/everyFrames/niko_MPC_temporarySatelliteFleetDespawner.kt
package data.scripts.everyFrames

import com.fs.starfarer.api.campaign.CampaignFleetAPI
import com.fs.starfarer.api.campaign.LocationAPI
import data.scripts.campaign.econ.conditions.defenseSatellite.handlers.niko_MPC_satelliteHandlerCore
import data.utilities.niko_MPC_debugUtils
import data.utilities.niko_MPC_fleetUtils.satelliteFleetDespawn
import data.utilities.niko_MPC_fleetUtils.setTemporaryFleetDespawner
import data.utilities.niko_MPC_satelliteUtils.getSatelliteBattleTracker

class niko_MPC_temporarySatelliteFleetDespawner(
    var fleet: CampaignFleetAPI,
    var handler: niko_MPC_satelliteHandlerCore?
) : niko_MPC_baseNikoScript() {

    var advanceTimeSinceStart = 0.0f
    var graceRuns = 0

    init {
        fleet.setTemporaryFleetDespawner(this)
    }

    override fun runWhilePaused(): Boolean {
        return false
    }

    override fun getPrimaryLocation(): LocationAPI? {
        return fleet.containingLocation
    }

    override fun advance(amount: Float) {
        advanceTimeSinceStart += amount
        if (handler == null) {
            niko_MPC_debugUtils.displayError("Null handler during $this advance")
            delete()
            return
        }
        if (fleet.battle == null && graceRuns <= 0) {
            getRidOfFleet() //this despawns the fleet, which calls the listener, which calls prepareforgc. its weird
            return
        } else {
            val tracker = getSatelliteBattleTracker() ?: return //this cant ever happen in campaign
            val battle = fleet.battle
            if (battle != null && !tracker.areSatellitesInvolvedInBattle(battle, handler!!)) { // sanity
                tracker.associateSatellitesWithBattle(battle, handler!!, battle.pickSide(fleet))
            }
        }
        //graceRuns--;
    }

    override fun delete(): Boolean {
        if (!super.delete()) return false

        fleet.setTemporaryFleetDespawner(null)
        getRidOfFleet()
        fleet.removeScript(this)

        return true
    }

    override fun startImpl() {
        fleet.addScript(this)
    }

    override fun stopImpl() {
        fleet.removeScript(this)
    }

    //arbitrary number
    private fun getRidOfFleet() {
        if (!fleet.isDespawning) {
            val vanish = advanceTimeSinceStart < 1 //arbitrary number
            fleet.satelliteFleetDespawn(vanish)
        }
    }
}
